This patch is required as the frequency fixup in nslu2_init does not
run sufficiently early in the boot sequence to take effect. Therefore
current mainline behaviour does not take into account the altered
timer frequency. Thus a fixup is added earlier in the boot sequence
which does lead to correct timing on the NSLU2.

Signed-off-by: Michael-Luke Jones <mlj28@cam.ac.uk>

Index: linux-2.6.21-rc4-arm/arch/arm/mach-ixp4xx/nslu2-setup.c
===================================================================
--- linux-2.6.21-rc4-arm.orig/arch/arm/mach-ixp4xx/nslu2-setup.c	2007-03-20 11:44:30.000000000 -0800
+++ linux-2.6.21-rc4-arm/arch/arm/mach-ixp4xx/nslu2-setup.c	2007-03-20 11:45:56.000000000 -0800
@@ -159,8 +160,6 @@
 
 static void __init nslu2_init(void)
 {
-	ixp4xx_timer_freq = NSLU2_FREQ;
-
 	ixp4xx_sys_init();
 
 	nslu2_flash_resource.start = IXP4XX_EXP_BUS_BASE(0);
@@ -178,11 +177,19 @@
 	platform_add_devices(nslu2_devices, ARRAY_SIZE(nslu2_devices));
 }
 
+static void __init nslu2_fixup(struct machine_desc *desc,
+                struct tag *tags, char **cmdline, struct meminfo *mi)
+{
+	/* The xtal on this machine is non-standard. */
+	ixp4xx_timer_freq = NSLU2_FREQ;
+}
+
 MACHINE_START(NSLU2, "Linksys NSLU2")
 	/* Maintainer: www.nslu2-linux.org */
 	.phys_io	= IXP4XX_PERIPHERAL_BASE_PHYS,
 	.io_pg_offst	= ((IXP4XX_PERIPHERAL_BASE_VIRT) >> 18) & 0xFFFC,
 	.boot_params	= 0x00000100,
+	.fixup          = nslu2_fixup,
 	.map_io		= ixp4xx_map_io,
 	.init_irq	= ixp4xx_init_irq,
 	.timer          = &ixp4xx_timer,
